README.md

```diff
@@ -14,16 +14,119 @@ open-feature = { git = "https://github.com/open-feature/rust-sdk", branch = "mai
 
-## Roadmap
+Let's roll.
+
+### Basic Usage
+
+```rust
+#[derive(Clone, Default, Debug)]
+struct MyStruct {}
+
+#[tokio::test]
+async fn example() {
+    // Acquire an OpenFeature API instance.
+    // Note the `await` call here because asynchronous lock is used to guarantee thread safety.
+    let mut api = OpenFeature::singleton_mut().await;
+
+    // Createa a global evaluation context and set it into the API.
+    // Note that this is optional. By default it uses an empty one.
+    let global_evaluation_context = EvaluationContext::default();
+    api.set_evaluation_context(global_evaluation_context).await;
+
+    // Set the default feature provider.
+    // If you do not do that, [`NoOpProvider`] will be used by default.
+    //
+    // By default, [`NoOpProvider`] will simply return the default value of each type.
+    // You can inject value you want via its builder or evaluation context. See its document
+    // for more details.
+    //
+    // If you set a new provider after creating some clients, the existing clients will pick up
+    // the new provider you just set.
+    api.set_provider(NoOpProvider::default()).await;
+
+    // Create an unnamed client.
+    let client = api.create_client();
+
+    // Createa an evaluation context.
+    // It supports types mentioned in the specification.
+    //
+    // You have multiple ways to add a custom field.
+    let evaluation_context = EvaluationContext::builder()
+        .targeting_key("Targeting")
+        .build()
+        .with_custom_field("bool_key", true)
+        .with_custom_field("int_key", 100)
+        .with_custom_field("float_key", 3.14)
+        .with_custom_field("string_key", "Hello".to_string())
+        .with_custom_field("datetime_key", time::OffsetDateTime::now_utc())
+        .with_custom_field(
+            "struct_key",
+            EvaluationContextFieldValue::Struct(Arc::new(MyStruct::default())),
+        )
+        .with_custom_field("another_struct_key", Arc::new(MyStruct::default()))
+        .with_custom_field(
+            "yet_another_struct_key",
+            EvaluationContextFieldValue::new_struct(MyStruct::default()),
+        );
+
+    assert_eq!(
+        client
+            .get_bool_value("key", Some(&evaluation_context), None)
+            .await
+            .unwrap(),
+        bool::default()
+    );
+
+    // Create a named provider and bind it.
+    api.set_named_provider("named", NoOpProvider::builder().int_value(42).build())
+        .await;
 
-## Pending Feature List
+    // This named client will use the feature provider bound to this name.
+    let client = api.create_named_client("named");
+
+    assert_eq!(42, client.get_int_value("key", None, None).await.unwrap());
+
+    // Let's get evaluation details.
+    // Note that we will inject `300` as the int value via evaluation context.
```

Comment:
This is surprising. Is this because you've opted not to allow for default arguments to be passed in?

Reply:
Yes. And this was added mainly for testing purpose. By default it returns the default value of each type and you can inject values in multiple ways.

Reply:
Can you explain how fallback values work?

Reply:
Take `get_int_value` as an example. The caller gets a `Result<i64, EvaluationError>`. It has to handle both successful and failed requests anyway, so the following code should work:

```rust
let flag_value = client.get_int_details("flag", None, None).await.unwrap_or(some_default_value);
```

The point here is that we leverage Rust built-in feature to handle the abnormal evaluation and leave the error handle to the caller.

Reply:
Let me add some more background here. Last week I had a Slack thread with Todd talking about leveraging specific/advanced language features to model the normal/abnormal evaluation cases.
https://cloud-native.slack.com/archives/C0344AANLA1/p1695111710764429

And that led to the following PR: https://github.com/open-feature/spec/pull/209

The advantage of returning a `Result` is that the caller can use anything in the [toolbox](https://doc.rust-lang.org/std/result/enum.Result.html) Rust provided, and we do not need to re-invent the wheels. :-)

Reply:
Resolving this conversation. Feel free to re-open it if you want to further discuss about it.

Reply:
Thanks, that makes sense.